csrc/wmi.c

```diff
@@ -0,0 +1,125 @@
+#define _WIN32_DCOM
+#include <Wbemidl.h>
+#include <stdbool.h>
+
+#define OS_VERSION_MAX_SIZE 128
+
+
+HRESULT init() {
+  HRESULT hres = CoInitializeEx(0, COINIT_MULTITHREADED);
+
+  if (!FAILED(hres)) {
+    hres = CoInitializeSecurity(
+      NULL,
+      -1,
+      NULL,
+      NULL,
+      RPC_C_AUTHN_LEVEL_DEFAULT,
+      RPC_C_IMP_LEVEL_IMPERSONATE,
+      NULL,
+      EOAC_NONE,
+      NULL);
+  }
+  return hres;
+}
+
+HRESULT createInstance(IWbemLocator** plocator) {
+  return CoCreateInstance(
+    &CLSID_WbemLocator,
+    0,
+    CLSCTX_INPROC_SERVER,
+    &IID_IWbemLocator, (LPVOID*)plocator);
+}
+
+HRESULT connectServer(IWbemLocator* plocator, IWbemServices** pservices) {
+  if (!plocator) return E_INVALIDARG;
+
+  HRESULT hres = plocator->lpVtbl->ConnectServer(
+    plocator,
+    SysAllocString(L"ROOT\\CIMV2"), // Object path of WMI namespace
+    NULL,                           // User name. NULL = current user
+    NULL,                           // User password. NULL = current
+    NULL,                           // Locale. NULL indicates current
+    0,                              // Security flags.
+    NULL,                           // Authority (for example, Kerberos)
+    NULL,                           // Context object
+    pservices);
+
+  if (!FAILED(hres)) {
+    hres = CoSetProxyBlanket(
+      (IUnknown*)*pservices,
+      RPC_C_AUTHN_WINNT,           // RPC_C_AUTHN_xxx
+      RPC_C_AUTHZ_NONE,            // RPC_C_AUTHZ_xxx
+      NULL,                        // Server principal name
+      RPC_C_AUTHN_LEVEL_CALL,      // RPC_C_AUTHN_LEVEL_xxx
+      RPC_C_IMP_LEVEL_IMPERSONATE, // RPC_C_IMP_LEVEL_xxx
+      NULL,                        // client identity
+      EOAC_NONE);                  // proxy capabilities
+
+    if (FAILED(hres)) {
+      *pservices->lpVtbl->Release(pservices);
+      *pservices = NULL;
+    }
+  }
+
+  return hres;
+}
+
+HRESULT query(IWbemServices* pservices, BSTR query, IEnumWbemClassObject** penumerator) {
+  if (!pservices) return E_INVALIDARG;
+
+  return pservices->lpVtbl->ExecQuery(
+    pservices,
+    SysAllocString(L"WQL"),
+    query,
+    WBEM_FLAG_FORWARD_ONLY | WBEM_FLAG_RETURN_IMMEDIATELY,
+    NULL,
+    penumerator);
+}
+
+HRESULT getStringField(IEnumWbemClassObject* penumerator, BSTR field, wchar_t* value) {
+  if (!penumerator) return E_INVALIDARG;
```

Comment:
This function doesn't have meaning if value == NULL

Reply:
This one needed even more changes.